internal/healthchecks/builtin.go

```diff
@@ -0,0 +1,216 @@
+package healthchecks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AreObjectsHealthy checks if the given resources are healthy.
+// It returns a nil error if all the resources are healthy, if any resource is not healthy, the error will
+// contain the GVK + namespace/resourceName and the error message of each unhealthy resource.
+//
+// The current list of supported resources is:
+// - Deployments
+// - StatefulSets
+// - DaemonSets
+// - ReplicaSets
+// - Pods
+// - APIServices
+// - CustomResourceDefinitions
+//
+// If the resource is not supported, it is assumed to be healthy.
+func AreObjectsHealthy(ctx context.Context, client client.Client, objects []client.Object) error {
+	var gvkErrors []error
+
+	// Create a new scheme to be able to convert the objects to their correct version.
+	s := runtime.NewScheme()
+	if err := apiregistrationv1.AddToScheme(s); err != nil {
+		return err
+	}
+	if err := apiextensionsv1.AddToScheme(s); err != nil {
+		return err
+	}
+	if err := appsv1.AddToScheme(s); err != nil {
+		return err
+	}
+	if err := corev1.AddToScheme(s); err != nil {
+		return err
+	}
+
+	for _, object := range objects {
+		// Convert the object to its proper type, deepCopy it to avoid mutating the original object.
+		obj, err := runtime.ObjectConvertor(s).ConvertToVersion(object.DeepCopyObject(), object.GetObjectKind().GroupVersionKind().GroupVersion())
+		if err != nil {
+			gvkErrors = appendResourceError(gvkErrors, object, err.Error())
+			continue
+		}
+		switch obj := obj.(type) {
+		case *appsv1.Deployment:
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			}, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.Conditions == nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, "Deployment has no conditions")
+				continue
+			}
+
+			// Check if the deployment is available.
+			conditionExists := false
+			for _, condition := range obj.Status.Conditions {
+				if condition.Type == appsv1.DeploymentAvailable {
+					if condition.Status != "True" {
+						gvkErrors = appendResourceError(gvkErrors, obj, condition.Message)
+					}
+					conditionExists = true
+					break
+				}
+			}
```

Comment:
Can we use this helper function? https://github.com/kubernetes/apimachinery/blob/82362bc34ae0cf9ca8f904f7968b6bb257b6fb3d/pkg/api/meta/conditions.go#L74

Reply:
Ah, right. I forgot that there are some core types that don't use the conventional Condition type.